sycl/doc/extensions/proposed/sycl_ext_oneapi_annotated_ptr.asciidoc

```diff
@@ -0,0 +1,510 @@
+
+    // Conversion function
+    template <class U, class P> explicit annotated_ptr(annotated_ptr<U, P> const &) noexcept;
+
+    reference operator*() const noexcept;
+    reference operator[](std::ptrdiff_t) const noexcept;
+    annotated_ptr operator+(unsigned long long) const noexcept;
+    std::ptrdiff_t operator-(annotated_ptr) const noexcept;
+
+    operator bool() const noexcept;
+
+    // Implicit conversion is not supported
+    operator T*() noexcept = delete;
+    operator const T*() const noexcept = delete;
+
+    T* get() noexcept;
+    const T* get() const noexcept;
+
+    annotated_ptr& operator=(const T*) noexcept;
+    annotated_ptr& operator=(annotated_ptr const&) noexcept;
+
+    annotated_ptr& operator++() noexcept;
+    annotated_ptr operator++(int) noexcept;
+    annotated_ptr& operator--() noexcept;
+    annotated_ptr operator--(int) noexcept;
+
+    template<typename propertyT>
+    static constexpr bool has_property();
+
+    // The return type is an unspecified internal class used to represent
+    // instances of propertyT
+    template<typename propertyT>
+    static constexpr /*unspecified*/ get_property();
+
+  private:
+    T *Ptr;
+  };
+} // namespace sycl::ext::oneapi
+
+----
+
+[frame="topbot",options="header"]
+|===
+|Functions |Description
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ptr() noexcept;
+----
+|
+Constructs an `annotated_ptr` object. Does not allocate new storage. The underlying pointer is initialized to `nullptr`.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+explicit annotated_ptr(T *Ptr);
+----
+|
+Constructs an `annotated_ptr` object. Does not allocate new storage. The underlying pointer is initialized with `Ptr`.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ptr(annotated_ptr const &) noexcept = default;
+----
+|
+Constructs an `annotated_ptr` object from another `annotated_ptr` with the same template parameterization object.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+template <class U, class P> explicit annotated_ptr(annotated_ptr<U, P> const & ConvertFrom);
+----
+|
+Constructs the `annotated_ptr` object from the `ConvertFrom` object if the template parameter types are compatible. For example, if the properties on one of the operands conflict with the others then the compiler should issue a compile time error.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+reference operator*() const;
+----
+|
+Returns a reference wrapper which can be used to read or write to the underlying pointer. Reads/Writes using the reference will retain the annotations.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+reference operator[](std::ptrdiff_t Index) const;
+----
+|
+Returns an `annotated_ref` reference wrapper to the object at offset `Index`.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ptr operator+(unsigned long long Offset) const;
+----
+|
+Returns an `annotated_ptr` that points to the location `Offset` distance away from the underlying pointer.
```

Comment:
This "'Offset' distance away" wording might be confusing.  Should we say 'Offset' elements away?  Or perhaps we should define it in closer terms to the implementation and say "Returns an 'annotated_ptr' that points to this[Offset].